### PR TITLE
microxs from mg flux and chain file

### PR DESCRIPTION
**Pull request by jbae11**
# Description

This function will create a custom MicroXS object from an MG flux spectrum and chain file. It still needs a lot of work and I'm just putting it out so folks can fix it and provide advice. This is in hopes of creating a standalone activation capability, which I'm currently testing.

Lots of credit to @shimwell for the help and advice.

# Checklist

- [x] I have performed a self-review of my own code
- [ ] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

**Comment by shimwell** [comment]
Congrats on your first PR @jbae11 I can tinker with this a bit and see what change be changed. Thanks for going the extra mile and converting the discussion to real code.

**Comment by shimwell** [comment]
@jbae11 I've made a PR to your branch that:

- changes the function to a class method

- adds some tests

The part I think needs some attention is the obtaining of the cross section for the MT number at the middle energy of the energy bin. I think it might be better to convert the cross section into a multi group cross section with the MGXS class and then get the xs value of the bin from the multigroup as this will apply some rules that take into account the shape of the cross section over the bin width.

**Comment by jbae11** [comment]
Thanks @shimwell ! I merged your edits. For the MGXS, I do think there's a better way than sampling at the midpoint - I'm not too clear on how to implement the MGXS, but if you have a good implementation feel free to make another MR :) 

One thing I'm keeping in mind is the speed at which this happens - ~3 minutes for an activation calculations is too long in my mind, so that's why I'm trying to keep everything a np array. Also found out that there are different [interpolation schemes available for the xs extraction] (https://github.com/openmc-dev/openmc/blob/develop/openmc/data/function.py#L184), so that's probably something we'll have to mess around once we get to validating.

**Comment by shimwell** [comment]
I shall look into the MGXS a bit more but I think it can help with bringing the time down as if I recall correctly we can make the cross sections in the desired group structure for the desired nuclides and export it to a h5 file ready for reuse. Now I'm wondering if we should also distribute multigroup cross sections in h5 format for a few popular multigroup formats

File: openmc/deplete/microxs.py
Comment (yardasol) [inline openmc/deplete/microxs.py:216]: What's the rationale for using `Sequence` instead of `List`?

File: openmc/deplete/microxs.py
Comment (yardasol) [inline openmc/deplete/microxs.py:221]:
```suggestion
        """Generated MicroXS object from a known flux and a chain file. The size of the MicroXS matrix depends
```

File: openmc/deplete/microxs.py
Comment (yardasol) [inline openmc/deplete/microxs.py:38]:
```suggestion
def _resolve_openmc_data_path(openmc_data_path: str=None):
```

File: openmc/deplete/microxs.py
Comment (yardasol) [inline openmc/deplete/microxs.py:236]:
```suggestion
        energies : iterable of float or str
            Energy group boundaries in [eV] or the name of the group structure
        multi_group_flux : iterable of float
            Energy-dependent multigroup flux values
        chain_file : str, optional
            Path to the depletion chain XML file that will be used in
            depletion simulation.
        openmc_data_path : str, optional
            Path to the cross section XML file that contains data library paths.
        temperature : int, optional
            Temperature for cross section evaluation in [K].
```

File: openmc/deplete/microxs.py
Comment (yardasol) [inline openmc/deplete/microxs.py:296]: I believe the second value in `microxs_arr` needs to be an element of `_valid_rxns` -- which is just the `chain.REACTIONS` variable with the addition of `'fission'` -- NOT an MT number, so we'll need to convert from MT back to internal OpenMC reaction names.

File: openmc/deplete/microxs.py
Comment (yardasol) [inline openmc/deplete/microxs.py:266]: Purpose for this?

File: openmc/deplete/microxs.py
Comment (yardasol) [inline openmc/deplete/microxs.py:219]:
```suggestion
        temperature: int=294
```

**Review by yardasol — CHANGES_REQUESTED** [review]
Hi @jbae11,
Thanks for this PR, I think this will be a good addition. I think we should have @paulromano take a look at this since he retrofitted the MicroXS class to support MGXS.

My other comments are below

File: openmc/deplete/microxs.py
Comment (shimwell) [reply openmc/deplete/microxs.py:216]: We use Sequence in a few other places in openmc and one advantage is that Sequence allows tuples and lists. This allows us to pass in non mutable tuples.

File: openmc/deplete/microxs.py
Comment (jbae11) [reply openmc/deplete/microxs.py:266]: probably not the best way to implent this, but since `fission` is not in the reactions variable, I'm adding it later but making sure that the order stays consistent by pushing fission (mt=18) at the end for both the `reactions` and `mt` variable.

File: openmc/deplete/microxs.py
Comment (jbae11) [reply openmc/deplete/microxs.py:296]: I might not be understanding properly, but it `mt_indx` should just be an integer right? For indexing the array? And that index should correspond to the `reactions`

Comment (yardasol) [reply openmc/deplete/microxs.py:296]: Ah, it seems I both misread and misunderstood what you were doing with the `microxs_arr`. You are absolutely correct.

**Comment by jbae11** [comment]
Currently, it takes a couple of minutes for this function to run, which I don't think is acceptable, given activation calculations are in the matter of seconds, is there a way to accelerate this?

**Comment by shimwell** [comment]
> Currently, it takes a couple of minutes for this function to run, which I don't think is acceptable, given activation calculations are in the matter of seconds, is there a way to accelerate this?

I took a look at the openmc.lib.nuclide.collapse_rate to see if this could help speed up the code but got stuck [here](https://github.com/openmc-dev/openmc/issues/2781)

**Comment by paulromano** [comment]
@jbae11 Thanks for the PR and so sorry I haven't had a chance to take a look until now! I'm glad that @shimwell brought up the idea of using `collapse_rate` because that's exactly the idea that I had in mind for such a functionality. I just commented on #2781 with a suggestion how to get it working, so I'll let you two take it from here but I think in general that is the right direction to take this PR. I think the workflow would look something like:
- Get the list of nuclides you need to get reaction rates for
- In a temporary directory, create a trivial OpenMC model with a single material containing all nuclides
- Call `openmc.lib.collapse_rate` for each nuclide/reaction with the appropriate flux spectrum (whole block surrounded by `openmc.lib.init` and `openmc.lib.finalize`)

**Comment by shimwell** [comment]
@jbae11 I've had a go at completing the suggestions from @paulromano and made a PR against your fork https://github.com/jbae11/openmc/pull/2 for your consideration

File: openmc/deplete/microxs.py
Comment (shimwell) [inline openmc/deplete/microxs.py:302]: I noticed this currently writes a summary.h5 file that takes a tiny amount of time and we don't need the file. So with this change we can avoid writing it :-)
```suggestion
            settings.output = {'summary': False}
```

**Comment by shimwell** [comment]
I've made an [example python script ](https://github.com/fusion-energy/neutronics-workshop/issues/257#issuecomment-1831667223)that does a depletion simulation with a pre calculated flux spectrum.

It does take a while to make a MicroXS file for all the isotopes found in a large chain file that also appear in the cross sections file. 

So I'm thinking this new class method could benefit from some more user control over how many nuclides are loaded up. Perhaps a chain_reduce argument can be added or the user could even be asked to specify the nuclides directly.

**Comment by shimwell** [comment]
Thinking a bit more about ways to speed up the micro generation and talking inspiration from the [get_microxs_and_flux ](https://docs.openmc.org/en/latest/pythonapi/generated/openmc.deplete.get_microxs_and_flux.html) method I plan to add a nuclides argument.

```
nuclides (list of str) – Nuclides to get cross sections for. If not specified, all burnable nuclides from the depletion chain file are used.
```

File: openmc/deplete/microxs.py
Comment (shimwell) [inline openmc/deplete/microxs.py:220]:
```suggestion
        temperature: int=294,
        nuclides: Optional[Iterable[str]] = None,
```

File: openmc/deplete/microxs.py
Comment (shimwell) [inline openmc/deplete/microxs.py:241]:
```suggestion
        nuclides : list of str
                Nuclides to get cross sections for. If not specified, all burnable
                nuclides from the depletion chain file are used.
```

File: openmc/deplete/microxs.py
Comment (shimwell) [inline openmc/deplete/microxs.py:268]:
```suggestion
        if not nuclides:
                nuclides = chain.nuclides
        reactions = chain.reactions
```

File: openmc/deplete/microxs.py
Comment (shimwell) [inline openmc/deplete/microxs.py:266]: This ```nuclides_with_data``` can be used later when adding nuclides to make sure we only add nuclides that we have data for
```suggestion
        nuclides_with_data = _get_nuclides_with_data(cross_sections)

```

File: openmc/deplete/microxs.py
Comment (shimwell) [inline openmc/deplete/microxs.py:311]: nuclides_with_data is defined earlier, this follows the methods used in ```get_microxs_and_flux``` a bit closer
```suggestion
                if nuc in nuclides_with_data:
```

File: openmc/deplete/microxs.py
Comment (shimwell) [inline openmc/deplete/microxs.py:265]: renaming to match names used in ```cross_sections ```
```suggestion
        cross_sections = openmc.data.DataLibrary.from_xml(openmc_data_path)
```

File: openmc/deplete/microxs.py
Comment (shimwell) [inline openmc/deplete/microxs.py:236]:
```suggestion
            depletion simulation.  Defaults to ``openmc.config['chain_file']``.
```

File: openmc/deplete/microxs.py
Comment (shimwell) [inline openmc/deplete/microxs.py:238]:
```suggestion
            Path to the cross section XML file that contains data library paths.
            Defaults to ``openmc.config['cross_sections']``.
```

**Comment by shimwell** [comment]
Sorry @jbae11 I've made another PR against your branch to fix some mistakes I introduced :flushed: 

This latest version allows us to optionally limit the micro_xs to user specified nuclides. This will produce a micro_xs quickly compared to the nuclides from the entire chain file (however this remains an option as well)

I have added another test to show how the nuclides option works

File: openmc/deplete/microxs.py
Comment (jbae11) [reply openmc/deplete/microxs.py:310]: yes! I believe so.

Comment (jbae11) [reply openmc/deplete/microxs.py:310]: so using this I get the very similar (not same due to the small differences from flux interpolation / data) microxs as ORIGEN

Comment (shimwell) [inline openmc/deplete/microxs.py:310]:
```suggestion
                        collapse = lib_nuc.collapse_rate(
                            c_mt, c_temperature, energies, norm_multi_group_flux
```

File: openmc/deplete/microxs.py
Comment (shimwell) [inline openmc/deplete/microxs.py:302]:
```suggestion
            norm_multi_group_flux = [val*1/sum(multi_group_flux) for val in multi_group_flux]
```

File: openmc/deplete/microxs.py
Comment (jbae11) [reply openmc/deplete/microxs.py:310]: lgtm!

Comment (shimwell) [reply openmc/deplete/microxs.py]: This is now normalised

**Comment by jbae11** [comment]
@yardasol have we adequately addressed your comments?

**Comment by paulromano** [comment]
@jbae11 @shimwell I've made some direct updates on this branch:
- Changed name to `from_multigroup_flux`
- Simplified logic in `from_multigroup_flux`, notably with how MT values are determined
- Adding an optional `reactions` argument that allows a user to specify which reactions they want

Let me know if either of you have any objections to these changes. If not, I think this should be good to merge!

**Review by yardasol — APPROVED** [review]
Hi @jbae11, these changes look good! I just have one question I'd like answered, it is below.

**Comment by shimwell** [comment]
I think @jbae11 has a few comparison simulations that might be good to run. @jbae11 is it easy to run your activator code and make [these ](https://github.com/fusion-energy/neutronics-workshop/issues/257#issuecomment-1855938576) graphs again?

File: openmc/deplete/microxs.py
Comment (paulromano) [reply openmc/deplete/microxs.py:315]: The flux is normalized in this method so that the sum of the values is one. However, now that I'm thinking about it, what we really want is a normalization such that $\sum\limits_g \int_{E_{g+1}}^{E_g} \phi(E) = 1$. Do others agree?

**Comment by jbae11** [comment]
Thanks @yardasol  @shimwell @paulromano !

@paulromano - I agree with your normalization comment. Also thanks for the great edits!
@yardasol Let me know if this completes your review.
@shimwell those graphs look the same, and the discrepancy is not from the microxs, but from the operator/integrator, since the microxs values correspond with SCALE/COUPLE (SCALE's collapsed microxs generating code) values.

**Comment by paulromano** [comment]
Ok, after thinking about the normalization more, dividing by the sum is fine as long as it is assumed that the flux values are absolute (not per eV). I added a test confirming that it works as intended.

**Comment by shimwell** [comment]
Thanks for the extra test and improvements @paulromano

Thanks @jbae11 for confirming that Microxs is good and results differ elsewhere (perhaps operator/integrator). I think we should keep working on that seperatly to this PR

This PR looks good to me, it does what it set out to do. I see it has an approval already let's merge this tomorrow unless there are any objections?

**Review by shimwell — APPROVED** [review]

**Review by paulromano — APPROVED** [review]

**Pull request merged**
